packages/block/dom.ts

```diff
@@ -0,0 +1,178 @@
+          const pool = el[EVENT_LISTENERS_POOL];
+          if (pool) {
+            const listeners = pool[event] as Record<string, any> | undefined;
+            if (listeners) {
+              const handlers = Object.values(listeners);
+              for (let i = 0, j = handlers.length; i < j; ++i) {
+                handlers[i](nativeEventObject);
+              }
+            }
+          }
+          el = el.parentNode;
+        }
+      },
+      {
+        capture: false,
+      },
+    );
+    document[EVENTS_REGISTRY].add(event);
+  }
+  const pointer = listenerPointer++;
+  if (mapHas$.call(eventCache, value)) return mapGet$.call(eventCache, value);
+  const patch = (newValue?: EventListener) => {
+    if (!el[EVENT_LISTENERS_POOL]) el[EVENT_LISTENERS_POOL] = {};
+    const pool = el[EVENT_LISTENERS_POOL];
+    if (!pool[event]) pool[event] = {};
+    pool[event][pointer] = (e: Event) => {
+      (newValue ?? value)?.(e);
+    };
+  };
+  patch(value);
+  mapSet$.call(eventCache, value, patch);
+  return patch;
+};
+
+export const insertText = (
+  el: HTMLElement,
+  value: any,
+  index: number,
+): Text => {
+  const node = document.createTextNode(value);
+  const childNodes = childNodes$.call(el);
+  if (hasChildNodes$.call(el) && index < childNodes.length) {
+    insertBefore$.call(el, node, childNodes[index]);
+  } else {
+    appendChild$.call(el, node);
+  }
+  return node;
+};
+
+export const setText = (el: HTMLElement, value: string, index: number) => {
+  setTextContent$.call(childNodes$.call(el)[index], value);
+};
+
+export const setAttribute = (
+  el: HTMLElement,
+  name: string,
+  value?: string | boolean | null,
+): void => {
+  const isValueNully = value === undefined || value === null;
+  value = isValueNully ? '' : value;
+  if (name === 'style') {
+    if (typeof value === 'string') {
+      el.style.cssText = value;
+    } else if (name.startsWith('-')) {
+      el.style.setProperty(name, String(value));
+    } else if (isValueNully) {
+      el.style[name] = '';
+    } else if (typeof value !== 'number' || IS_NON_DIMENSIONAL.test(name)) {
+      el.style[name] = value;
+    } else {
+      el.style[name] = `${String(value)}px`;
+    }
+    return;
+  }
+  // Handle SVG namespaced attributes
+  if (name.charCodeAt(0) === X_CHAR) {
+    // eslint-disable-next-line prefer-named-capture-group
+    name = name.replace(/xlink(H|:h)/, 'h').replace(/sName$/, 's');
+    if (name.startsWith('xmlns')) {
+      setAttributeNS$.call(el, XML_NS, name, String(value));
+    } else if (name.startsWith('xlink')) {
+      setAttributeNS$.call(el, XLINK_NS, 'href', String(value));
+    }
+    return;
+  }
+  if (
+    el[name] !== undefined &&
+    el[name] !== null &&
+    !Reflect.has(el.style, name) &&
+    !(el instanceof SVGElement) &&
+    name !== 'href' &&
+    name !== 'list' &&
+    name !== 'form' &&
+    // cast to `0` instead
+    name !== 'tabIndex' &&
+    name !== 'download' &&
+    name in el
+  ) {
```

Comment:
This whole condition can be faster by using `!Set#has(name)` and moving `name in el` as the very beginning (as a fast path)

Reply:
And another issue. This condition uses both `!Reflect.has(el.style, name)` and `name in el`. Ideally, this should be consistent (choose only one of them).

Reply:
Fixed, list of props hoisted to a set, and chose `name in el` over Reflect